### PR TITLE
Core/Scripts: Add Shade of Aran & Atiesh interaction

### PR DESCRIPTION
**Pull request by Krudor**
[//]: # (***************************)
[//]: # (** FILL IN THIS TEMPLATE **)
[//]: # (***************************)

**Changes proposed:**

- Implement Shade of Aran & Atiesh interaction

**Target branch(es):** 3.3.5/master

**Tests performed:** Builds, in-game tested

[YouTube - Video of said interaction](https://www.youtube.com/watch?v=_PSF0VBJOBY)

### CODE REVIEW

**Comment by Krudor** [comment]
Forgot to squash my commit history, doh

**Comment by Keader** [comment]
You can squash your commits in any moment

```git rebase -i HEAD~2``` where "2" is the number of commits that you desire squash
change commits with fix for ```fixup```, and force push after squash.

btw, you should use one branch for each PR, avoid this "commit" flood

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp [hunk 1]
```diff
@@ -79,6 +80,23 @@ enum SuperSpell
...
 
+enum Items
+{
+    ITEM_ATIESH_MAGE            = 22589,
+    ITEM_ATIESH_WARLOCK         = 22630,
+    ITEM_ATIESH_PRIEST          = 22631,
+    ITEM_ATIESH_DRUID           = 22632,
+};
+
+uint32 const AtieshStavesCnt = 4;
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: you realy need this constantant?
you only use it in ```AtieshStaves[AtieshStavesCnt]```, you can put directly ```AtieshStaves[4]```

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp [hunk 2]
```diff
@@ -97,6 +115,13 @@ class boss_shade_of_aran : public CreatureScript
             instance = creature->GetInstanceScript();
         }
 
+        void InitializeAI() override
+        {
+            ScriptedAI::InitializeAI();
+
+            SeenAtiesh = false;
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: this override is realy need?
you can initialize this variable in ```void Initialize()``` function

Comment (Keader) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: this override is realy need?
you can initialize this variable in ```void Initialize()``` function

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp [hunk 3]
```diff
@@ -503,6 +529,31 @@ class boss_shade_of_aran : public CreatureScript
...
+            ScriptedAI::MoveInLineOfSight(who);
+
+            if (SeenAtiesh || me->IsInCombat() || me->GetDistance2d(who) > me->GetAttackDistance(who) + 10.0f)
+                return;
+
+            Player* player = who->ToPlayer();
+            if (!player)
+                return;
+
+            for (auto id : AtieshStaves)
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: i saw someone complain about this "auto" in another commit
why not use ```for (uint8 i = 0; i < 4; ++i) ``` ?

Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: easyer to use auto : {} insted of for  i = 0; i < 4; ++i

Comment (Keader) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: true, but in source has more cases like: ```for (uint8 i = 0; i < 4; ++i)```
i think Jackpoz that complain about this "auto", idk what is real difference

Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: can be just for (auto const& i : {ITEM_ATIESH_MAGE, ITEM_ATIESH_WARLOCK, ITEM_ATIESH_PRIEST, ITEM_ATIESH_DRUID})   and it'll take less lines of code - faster to write it atleast, also'll not cause warnings and other shit

**Comment by Rushor** [comment]
@jackpoz you already closed https://github.com/TrinityCore/TrinityCore/issues/12918. do you think it would be fair to close this pr too?

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp (hunk 3 above)
Comment (Shauren) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: We use auto only for iterator types (container<T>::iterator)

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp [hunk 4]
```diff
@@ -503,6 +529,31 @@ class boss_shade_of_aran : public CreatureScript
...
+            {
+                if (!player->HasItemOrGemWithIdEquipped(id, 1))
```
Comment (Shauren) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: HasItemOrGemWithIdEquipped has to iterate all equipment slots and their gems, wouldnt it be better to just take the main hand weapon and check the entry?

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp (hunk 3 above)
Comment (HelloKitty) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: The C++11 range-based for doesn't require auto, the Type name is also small in this case. Use the enum Type name or uint8 like @Tonghost said.

**Comment by Shauren** [comment]
@Rushor u wot?

**Comment by tkrokli** [comment]
I think Rushor meant it as a rhetorical question, because jackpoz closed #12918 as "too cosmetic/won't fix".

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp (hunk 1 above)
Comment (Krudor) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: Of course we don't need it, but this is how I want to write it.

Comment (Krudor) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: Actually, I was planning to use this variable for my weapon iteration loop, but seeing as I'm not using it anyway, might as well remove it.

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp [hunk 5]
```diff
@@ -541,9 +540,9 @@ class boss_shade_of_aran : public CreatureScript
             if (!player)
                 return;
 
-            for (auto id : AtieshStaves)
+            for (uint8 id : AtieshStaves)
```
Comment (Shauren) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: uint8? really? whats the type of AtieshStaves[0]?

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp [hunk 6]
```diff
@@ -554,6 +553,19 @@ class boss_shade_of_aran : public CreatureScript
                 break;
             }
         }
+
+        private:
+            bool PlayerHasWeaponEquipped(Player* player, uint32 itemEntry)
+            {
+                for (uint8 slot : {EQUIPMENT_SLOT_MAINHAND, EQUIPMENT_SLOT_OFFHAND, EQUIPMENT_SLOT_RANGED})
```
Comment (Shauren) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: Let's not play custom here, staves are 2handed and go into mainhand only.

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp (hunk 5 above)
Comment (Krudor) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp]: Shit...

**Pull request closed without merging**